src/bridge/cockpitrouter.c

```diff
@@ -1188,6 +1188,9 @@ void cockpit_router_set_bridges (CockpitRouter *self,
       rule = l->data;
       if (rule->config)
         {
+          if (rule == self->superuser_rule)
+            self->superuser_rule = NULL;
```

Comment:
Rules for running peers are supposed to stay intact, no?  See `cockpit_router_set_bridges`, which has explicit code to move a old rule over to the new list intact.  That should continue to happen, I'd say.

Does `router_rule_find` maybe just not find the running bridges at all anymore and kills them all?

Setting `superuser_rule` to NULL will block future superuser channels.  The router behaves as if no superuser bridge would be running.

Reply:
This code should "never happen".  I only stumbled upon it because I accidentally broke the exact check that you're talking about.  That's fixed now, so this is back to being more or less dead code (unless for some reason the on-disk manifests actually changed).  But in that (rare) case, this will lead to a definite crash, so we should absolutely fix it.

Reply:
Ahh, this is for the case where the superuser bridge config is actually removed from the manifests for real, and we really want to kill it. This is indeed broken on master.
I think there might be more things to do to cleanly stop the superuser peer. I'll have a look.

Reply:
An easy way to produce the crash with the current version of cockpit:

- login as admin, with admin access on
- `sed -e 's/sudo/false/' -i /usr/share/cockpit/shell/manifest.json`
- refresh the page

for me this will crash every time because of the sudo rule being replaced and the dangling pointer not being cleared.

Reply:
There is no "closed" event on the superuser transport when the CockpitPeer is destroyed by unreffing it, we need to do this, I guess:

```
          if (self->superuser_rule == rule)
            on_superuser_transport_closed (self->superuser_transport, "terminated", self);
```

Reply:
But just setting superuser_rule = NULL is more obviously correct so I am happy with that.  Uff, how difficult can I be? :)